source/detail/parsers.cpp

```diff
@@ -0,0 +1,766 @@
+            if (errno_ref == ERANGE)
+            {
+                ok = false;
+            }
+
+            if (ok)
+            {
+                result = parsed;
+            }
+        }
+    }
+
+    return ok;
+}
+
+bool xlnt::detail::parse(const std::string &string, unsigned long long &result, std::size_t *num_characters_parsed, int base)
+{
+    errno = 0; // Reset errno; note: since c++11, errno is thread-local and thus safe
+    bool ok = !string.empty();
+
+    if (ok)
+    {
+        char *end = nullptr;
+        ok = parse(string.c_str(), result, &end, base);
+
+        if (num_characters_parsed != nullptr)
+        {
+            *num_characters_parsed = end - string.c_str();
+        }
+    }
+    else if (num_characters_parsed != nullptr)
+    {
+        *num_characters_parsed = 0;
+    }
+
+    return ok;
+}
+
+bool xlnt::detail::parse(const std::string &string, unsigned int &result, std::size_t *num_characters_parsed, int base)
+{
+    errno = 0; // Reset errno; note: since c++11, errno is thread-local and thus safe
+    bool ok = !string.empty();
+
+    if (ok)
+    {
+        char *end = nullptr;
+        ok = parse(string.c_str(), result, &end, base);
+
+        if (num_characters_parsed != nullptr)
+        {
+            *num_characters_parsed = end - string.c_str();
+        }
+    }
+    else if (num_characters_parsed != nullptr)
+    {
+        *num_characters_parsed = 0;
+    }
+
+    return ok;
+}
+
+bool xlnt::detail::parse(const std::string &string, unsigned long &result, std::size_t *num_characters_parsed, int base)
+{
+    errno = 0; // Reset errno; note: since c++11, errno is thread-local and thus safe
+    bool ok = !string.empty();
+
+    if (ok)
+    {
+        char *end = nullptr;
+        ok = parse(string.c_str(), result, &end, base);
+
+        if (num_characters_parsed != nullptr)
+        {
+            *num_characters_parsed = end - string.c_str();
+        }
+    }
+    else if (num_characters_parsed != nullptr)
+    {
+        *num_characters_parsed = 0;
+    }
+
+    return ok;
+}
+
+bool xlnt::detail::parse(const char *string, double &result, char **end)
+{
+    return parse(detail::get_system_locale(), string, result, end);
+}
+
+bool xlnt::detail::parse(const char *string, float &result, char **end)
+{
+    return parse(detail::get_system_locale(), string, result, end);
+}
+
+bool xlnt::detail::parse(const char *string, long double &result, char **end)
+{
+    return parse(detail::get_system_locale(), string, result, end);
+}
+
+// The following function must be "static" to avoid noexcept-type warnings / errors on GCC. See https://stackoverflow.com/a/46857525
+template <typename T, typename ParseFunc>
+static bool parse_number(const std::locale &loc, const char *string, T &result, char **end, ParseFunc func)
+{
+    // Only accept float, double and long double, but NOT any other floating-point types (e.g. extended floating-point types)
```

Comment:
Oh, I see, you are right. Let me explain my confusion: I did, in fact, already have to deal with numbers that have other decimal separators. They were in cells of type `xlnt::cell_type::formula_string` and were the result of formulas that have been cached by Excel. In other words, Excel saves not only the formula, but also caches the last result of the formula. Since my project did not actually compute the result of formulas, but was simply trying to read it from Excel (and thus XLNT), I had to parse the string, which contained the result exactly as shown to the user, of course in the user's locale. This means that I really needed to have a very robust parser that is able to parse more than just `.` as the decimal separator.

However, looking at the Office Open XML specification, apart from the exception I mentioned, it does seem to always serialize numbers as if the "C" locale (`std::locale::classic()`) was used - that is, without thousands separators and with `.` as the decimal separator. However, for `formula_strings`, XLNT does NOT parse anything by itself, but simply treats it as a string. I think that the parsing logic I explained above is then rather application-specific and should not be specifically handled by XLNT.

In other words, I agree with you and propose the following to ensure maximum performance:

1. Parse the number as it is.
2. If we stopped at the dot `.`, copy the string then replace only that character (by calculating the index using the difference to the beginning pointer) with the locale's decimal separator, then parse again.

This way we should be able to achieve the maximum possible performance with the least amount of unnecessary copying and/or replacements. If you agree to this, let me know and I'll change it :wink:

Reply:
Maybe we can keep your current implementation somewhere in case we later wan't to assist the user to parse the cached results.

Reply:
> Maybe we can keep your current implementation somewhere in case we later wan't to assist the user to parse the cached results.

To avoid code duplication, I could add an additional parameter for this case, which avoids additional parsing by default.


> Is this correct if "." is used as a thousand separator in the current locale?

Good question - I added some unit tests for this case and they run fine. I'll check even further whether this could be an issue.

> Is this faster than first looking for a "." and replacing it with the locale decimal separator if found?

While I wanted to avoid unnecessary copying at all costs, if we know that replacements are necessary, it should be even faster if we copy the string and simply replace the first separator we encounter (we shouldn't need to replace everything as long as the thousands separators are not an issue, in which case we have bigger issues anyway...).